build/tools/CFamilyJarPreProcessor/Preprocessor.cs

```diff
@@ -0,0 +1,124 @@
+        private readonly string[] SingleFilePatterns = new string[]
+        {
+                "Sonar_way_profile.json",
+                "RulesList.json",
+                TarUnzipSubFolder + @"\subprocess.exe",
+                TarUnzipSubFolder + @"\LICENSE_THIRD_PARTY.txt"
+        };
+
+        // List of patterns to match multiple files in the uncompressed output
+        private readonly string[] MultipleFilesPatterns = new string[]
+        {
+            @"org\sonar\l10n\cpp\rules\params\*_params.json",
+            @"org\sonar\l10n\cpp\rules\cpp\*.json"
+        };
+
+        private readonly ILogger logger;
+        public Preprocessor(ILogger logger)
+        {
+            this.logger = logger;
+        }
+
+        /// <summary>
+        /// Downloads plugin, extracts files, and copies the requires file to the output folder 
+        /// </summary>
+        /// <remarks>The method is lazy i.e. won't download the plugin if it exists, won't copy a file
+        /// unless it is newer than the target file.</remarks>
+        public void Execute(string downloadUrl, string destinationDir)
+        {
+            var pluginFileName = Common.ExtractPluginFileNameFromUrl(downloadUrl, logger);
+
+            // Ensure working directories exists
+            var localWorkingFolder = Common.GetLocalBuildTimePluginCacheDir("SLVS_CFamily_Build");
+            var perVersionPluginFolder = Path.Combine(localWorkingFolder, Path.GetFileNameWithoutExtension(pluginFileName));
+            Common.EnsureWorkingDirectoryExist(perVersionPluginFolder, logger);
+
+            // Download and unzip the jar
+            var jarFilePath = Path.Combine(perVersionPluginFolder, pluginFileName);
+            Common.DownloadJarFile(downloadUrl, jarFilePath, logger);
+            Common.UnzipJar(jarFilePath, perVersionPluginFolder, logger);
+
+            // Uncompress and extract the windows tar archive to get the subprocess exe
+            var tarFilePath = Common.FindSingleFile(perVersionPluginFolder, WindowsTxzFilePattern, logger);
+            var tarSubFolder = Path.Combine(perVersionPluginFolder, TarUnzipSubFolder);
+            Common.UncompressAndUnzipTgx(tarFilePath, tarSubFolder, logger);
+
+            // Locate the required files from the uncompressed jar and tar
+            var fileList = FindFiles(perVersionPluginFolder);
+
+            // Copy the files to the output directory
+            CopyFilesToOutputDirectory(fileList, destinationDir);
+        }
+
+        private List<string> FindFiles(string searchRoot)
+        {
+            var files = new List<string>();
+
+            files.AddRange(Common.FindSingleFiles(searchRoot, SingleFilePatterns, logger));
+            files.AddRange(Common.FindMultipleFiles(searchRoot, MultipleFilesPatterns, logger));
+
+            return files;
+        }
+
+        private void CopyFilesToOutputDirectory(IList<string> files, string destinationDir)
```

Comment:
This method and the one below are new.

Logically, the net effect of calling the task is still the same - the jar is downloaded and the files are copied to the destination folder.

However, in the MSBuild Task version, the task return the list of files to the MSBuild target which then did the copy.

Now, the new exe does the copy part too i.e. this method and the one below.